### PR TITLE
KAFKA-9000: fix flaky FK join test by using TTD

### PR DESCRIPTION
**Pull request by vvcephei**
Migrate this integration test to use TopologyTestDriver instead of running 3 Streams instances.

Dropped one test that was attempting to produce specific interleavings. If anything, these should be verified deterministically by unit testing.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Checkstyle error:
```
[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.12/streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:66:37: Name 'streamsConfig' must match pattern '(^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$)|(^log$)'. [ConstantName]
```

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 1]
```diff
@@ -27,673 +25,314 @@
...
+import static java.util.Collections.emptyMap;
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.apache.kafka.common.utils.Utils.mkProperties;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
 
 
-@Category({IntegrationTest.class})
+@Category(IntegrationTest.class)
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Is this still an integration test?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 2]
```diff
@@ -27,673 +25,314 @@
...
+                        is(expected)
+                    );
+                    assertThat(
+                        "leftJoin:" + leftJoin,
+                        asMap(store),
+                        is(expected)
+                    );
+                }
+
+                //Now delete one LHS entity such that one delete is propagated down to the output.
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: nit: space, ditto below.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 3]
```diff
@@ -27,673 +25,314 @@
...
+                final TestOutputTopic<Integer, String> outputTopic = driver.createOutputTopic(OUTPUT, new IntegerDeserializer(), new StringDeserializer());
+                final KeyValueStore<Integer, String> store = driver.getKeyValueStore("store");
+
+                right.pipeInput("1", 10L);
+                right.pipeInput("2", 20L);
+
+                left.pipeInput(1, 1.33f);
+                left.pipeInput(2, 2.77f);
+
+                {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Why adding the nested bracket? Seems unnecessary.

**Review by guozhangwang — APPROVED** [review]
LGTM. Also cc @bellemare for another look. 

This is semi-orthogonal, but there's a reported duplicate scenario on the original PR and I'm wondering if we can easily reproduce it in the unit test.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 4]
```diff
@@ -27,673 +25,314 @@
...
-    private static final  Properties CONSUMER_CONFIG = new Properties();
-    private static final Properties LEFT_PROD_CONF = new Properties();
-    private static final Properties RIGHT_PROD_CONF = new Properties();
+    private static final String LEFT_TABLE = "left_table";
+    private static final String RIGHT_TABLE = "right_table";
+    private static final String OUTPUT = "output-topic";
+    private static final Properties STREAMS_CONFIG = mkProperties(mkMap(
+        mkEntry(StreamsConfig.APPLICATION_ID_CONFIG, "ktable-ktable-joinOnForeignKey"),
+        mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "asdf:0000"),
+        mkEntry(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest"),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: we don't need this config for TTD

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 5]
```diff
@@ -27,673 +25,314 @@
...
+        mkEntry(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath()),
+        mkEntry(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, "0"),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: we don't need this config for TTD

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 6]
```diff
@@ -27,673 +25,314 @@
...
+        mkEntry(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, "100"),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: we don't need this config for TTD

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 7]
```diff
@@ -27,673 +25,314 @@
...
+        mkEntry(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Do we need/want this?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 8]
```diff
@@ -27,673 +25,314 @@
...
-                OUTPUT,
-                expected3.size());
-        assertEquals(expected3, result3);
-
-        //Ensure the state stores have the correct values within:
-        final Set<KeyValue<Integer, String>> expMatResults = new HashSet<>();
-        expMatResults.add(new KeyValue<>(1, "value1=1.11,value2=10"));
-        validateQueryableStoresContainExpectedKeyValues(expMatResults, currentMethodName);
+    public void doJoinFromLeftThenDeleteLeftEntity() {
+        for (final Boolean leftJoin : new Boolean[] {false, true}) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Reusing the code is fine, but I would not do it in a loop, but with two proper test methods (with corresponding names) and a single private method that is called by both passing int `leftJoin` as parameter.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 9]
```diff
@@ -27,673 +25,314 @@
...
-        expectedTwo.add(new KeyValue<>(3, "value1=1.11,value2=10"));
-        assertArrayEquals(resultTwo.toArray(), expectedTwo.toArray());
-
-        //Ensure the state stores have the correct values within:
-        final Set<KeyValue<Integer, String>> expMatResults = new HashSet<>();
-        expMatResults.addAll(expected);
-        expMatResults.addAll(expectedTwo);
-        validateQueryableStoresContainExpectedKeyValues(expMatResults, currentMethodName);
+    public void doJoinFromRightThenDeleteRightEntity() {
+        for (final Boolean leftJoin : new Boolean[] {false, true}) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: as above

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 10]
```diff
@@ -27,673 +25,314 @@
...
-                    .withValueSerde(Serdes.String())
-                    .withCachingDisabled();
-        } else {
-            throw new RuntimeException("Current implementation of join on foreign key requires a materialized store");
-        }
+        final Materialized<Integer, String, KeyValueStore<Bytes, byte[]>> materialized =
+            Materialized.<Integer, String>as(Stores.inMemoryKeyValueStore(queryableStoreName))
+                .withKeySerde(Serdes.Integer())
+                .withValueSerde(Serdes.String())
+                .withCachingDisabled();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: Should we test with caching (to make sure we execute the caching code path)? Note that TDD flushes after each `pipeInput()` anyway.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 11]
```diff
@@ -27,673 +25,314 @@
...
+
+        final Function<Float, String> extractor = value -> Integer.toString((int) value.floatValue());
+        final ValueJoiner<Float, Long, String> joiner = (value1, value2) -> "value1=" + value1 + ",value2=" + value2;
 
         if (leftJoin)
-            left.leftJoin(right, tableOneKeyExtractor, joiner, Named.as("customName"), materialized)
+            left.leftJoin(right,
+                          extractor,
+                          joiner,
+                          Named.as("customName"),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Do we need to name it?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 12]
```diff
@@ -27,673 +25,314 @@
...
+                          materialized)
                 .toStream()
                 .to(OUTPUT, Produced.with(Serdes.Integer(), Serdes.String()));
         else
-            left.join(right, tableOneKeyExtractor, joiner, materialized)
+            left.join(right,
+                      extractor,
+                      joiner,
+                      Named.as("customName"),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: as above

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 13]
```diff
@@ -27,673 +25,314 @@
...
+                final TestInputTopic<String, Long> right = driver.createInputTopic(RIGHT_TABLE, new StringSerializer(), new LongSerializer());
+                final TestInputTopic<Integer, Float> left = driver.createInputTopic(LEFT_TABLE, new IntegerSerializer(), new FloatSerializer());
+                final TestOutputTopic<Integer, String> outputTopic = driver.createOutputTopic(OUTPUT, new IntegerDeserializer(), new StringDeserializer());
+                final KeyValueStore<Integer, String> store = driver.getKeyValueStore("store");
+
+                right.pipeInput("1", 10L);
+                right.pipeInput("2", 20L);
+
+                left.pipeInput(1, 1.33f);
+                left.pipeInput(2, 2.77f);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: nit: can we use different keys? My first impression was that it's a primary key-join converting integer to string -- also, it's rather weird to use float and convert to int to string -- why such a complicated setup?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 14]
```diff
@@ -27,673 +25,314 @@
...
-                CONSUMER_CONFIG,
-                OUTPUT,
-                expected3.size());
-        assertEquals(expected3, result3);
-
-        //Ensure the state stores have the correct values within:
-        final Set<KeyValue<Integer, String>> expMatResults = new HashSet<>();
-        expMatResults.add(new KeyValue<>(1, "value1=1.11,value2=10"));
-        validateQueryableStoresContainExpectedKeyValues(expMatResults, currentMethodName);
+    public void doJoinFromLeftThenDeleteLeftEntity() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:87]: nit: `shouldJoinAndDeleteFromLeftKTable`

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 15]
```diff
@@ -27,673 +25,314 @@
...
-        final List<KeyValue<Integer, String>> expectedTwo = new LinkedList<>();
-        expectedTwo.add(new KeyValue<>(3, "value1=1.11,value2=10"));
-        assertArrayEquals(resultTwo.toArray(), expectedTwo.toArray());
-
-        //Ensure the state stores have the correct values within:
-        final Set<KeyValue<Integer, String>> expMatResults = new HashSet<>();
-        expMatResults.addAll(expected);
-        expMatResults.addAll(expectedTwo);
-        validateQueryableStoresContainExpectedKeyValues(expMatResults, currentMethodName);
+    public void doJoinFromRightThenDeleteRightEntity() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:165]: `shouldJoinAndDeleteFromRightTable`

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 13 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: To make the test even better, we should include one non-joining record and one actual 1:n join?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 16]
```diff
@@ -27,673 +25,314 @@
...
+                final TestOutputTopic<Integer, String> outputTopic = driver.createOutputTopic(OUTPUT, new IntegerDeserializer(), new StringDeserializer());
+                final KeyValueStore<Integer, String> store = driver.getKeyValueStore("store");
+
+                left.pipeInput(1, 1.33f);
+                left.pipeInput(2, 1.77f);
+                left.pipeInput(3, 3.77f);
+
+                right.pipeInput("1", 10L);
+                right.pipeInput("2", 20L);
+                right.pipeInput("3", 30L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: as above -- use different keys -- add non-joining record

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 17]
```diff
@@ -27,673 +25,314 @@
...
+            final Topology topology = getTopology(STREAMS_CONFIG, "store", leftJoin);
+            try (final TopologyTestDriver driver = new TopologyTestDriver(topology, STREAMS_CONFIG)) {
+                final TestInputTopic<String, Long> right = driver.createInputTopic(RIGHT_TABLE, new StringSerializer(), new LongSerializer());
+                final TestInputTopic<Integer, Float> left = driver.createInputTopic(LEFT_TABLE, new IntegerSerializer(), new FloatSerializer());
+                final TestOutputTopic<Integer, String> outputTopic = driver.createOutputTopic(OUTPUT, new IntegerDeserializer(), new StringDeserializer());
+                final KeyValueStore<Integer, String> store = driver.getKeyValueStore("store");
+
+                left.pipeInput(1, 1.33f);
+                left.pipeInput(2, 1.77f);
+                left.pipeInput(3, 3.77f);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: as above -- extend value for non-joining and 1:n joining case

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 18]
```diff
@@ -27,673 +25,314 @@
...
+        for (final Boolean leftJoin : new Boolean[] {false, true}) {
+            final Topology topology = getTopology(STREAMS_CONFIG, "store", leftJoin);
+            try (final TopologyTestDriver driver = new TopologyTestDriver(topology, STREAMS_CONFIG)) {
+                final TestInputTopic<String, Long> right = driver.createInputTopic(RIGHT_TABLE, new StringSerializer(), new LongSerializer());
+                final TestInputTopic<Integer, Float> left = driver.createInputTopic(LEFT_TABLE, new IntegerSerializer(), new FloatSerializer());
+                final TestOutputTopic<Integer, String> outputTopic = driver.createOutputTopic(OUTPUT, new IntegerDeserializer(), new StringDeserializer());
+                final KeyValueStore<Integer, String> store = driver.getKeyValueStore("store");
+
+                right.pipeInput("1", 10L);
+                right.pipeInput("2", 20L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: we should also put a non-joining entry into the right hand side table

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 19]
```diff
@@ -27,673 +25,314 @@
...
+                    assertThat(
+                        "leftJoin:" + leftJoin,
+                        asMap(store),
+                        is(expected)
+                    );
+                }
+
+                //Now delete one LHS entity such that one delete is propagated down to the output.
+
+                left.pipeInput(1, null);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: I think we should delete each record and verify directly after a single delete

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 20]
```diff
@@ -27,673 +25,314 @@
...
+                    );
+                    assertThat(
+                        "leftJoin:" + leftJoin,
+                        asMap(store),
+                        is(expected)
+                    );
+                }
+
+                //Now delete the RHS entity such that all matching keys have deletes propagated.
+                right.pipeInput("1", null);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: as above -- delete each record

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java [hunk 21]
```diff
@@ -27,673 +25,314 @@
...
-                QueryableStoreTypes.keyValueStore());
-
-        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreThree = streamsThree.store(queryableStoreName,
-                QueryableStoreTypes.keyValueStore());
-
-        // store only keeps last set of values, not entire stream of value changes
-        final Map<Integer, String> expectedInStore = new HashMap<>();
-        for (final KeyValue<Integer, String> expected : expectedResult) {
-            expectedInStore.put(expected.key, expected.value);
+    public void joinShouldProduceNullsWhenValueHasNonMatchingForeignKey() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: I think we should include this test in the two from above and remove this test method -- it's not really a special case that needs it's own test method IMHO.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Overall looks good. Couple of suggestions.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: I'd argue "yes", because this is not testing a "unit", like a processor or serde in isolation, but rather testing all the components of a FK join, as well as source and sink nodes, and the state stores in conjunction.

But I'd also argue that it doesn't matter what we call it :) . Do you prefer to drop this annotation for semantic reasons, or is there some practical difference in how these tests get run?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 3 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: It creates a scope, so we can be sure that we don't accidentally reference `expected` later on. This helps us avoid silly variable names like `expected1`, `expected2`, etc., as well as avoid the risk of accidentally referencing the wrong one.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 4 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: good call.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Not sure... Maybe it helps verify that the optimizer doesn't mess up the FK join?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: yeah, I think it's an artifact. I'll remove it.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 11 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: not sure why this was there.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 13 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Sure; I also found it a little confusing, but didn't want to just wholesale write a completely different test. I can do it, though.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 14 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:87]: I left these names on purpose because they are describing the actions taken, not the result. If you want it called "should", it would be something like "shouldWorkWhenJoiningFromTheLeftAndThenDeletingLeftEntity", which seems vapid. It seemed fine to just describe these tests in terms of what they are doing, since it would be obvious that they're then verifying the result is correct.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 19 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Didn't follow this. You're saying we should delete all the records instead of just one? Would that result in additional coverage?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 21 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Hmm. Not sure I agree. This case actually has its own top-level section in the KIP, namely what happens when you migrate from one non-existent FK to another. It seems worth explicitly calling out the truth table for this case.

The other tests are basically like, "what do I get if I add/remove a left record vs. a right record"? Sure, we could test all the permutations in one long test, but it's kind of hard to comprehend what such a test is doing after the fact.

**Comment by vvcephei** [comment]
@guozhangwang , I _think_ that reported duplicate scenario is either what is covered in `joinShouldProduceNullsWhenValueHasNonMatchingForeignKey` or the "rapidly changing FK" test I removed. I'm not entirely comfortable with removing the latter test, so maybe I'll add a unit test to exercise this functionality, since it's not possible with TTD.

**Comment by vvcephei** [comment]
Thanks for the reviews, all, it's a lot of good food for thought. I'll make some revisions and then ping you.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: How long does it take to complete one suite now? Usually if it is taking longer than a few tens of seconds, it should be integration test, otherwise unit test is fine. But this is not a very strict rule either.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Sounds good.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: ok. it takes 2 seconds on my computer. I'll drop this annotation.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: Huh. It looks like removing this results in not seeing all the tombstones we expect. I'll leave it in.

**Comment by guozhangwang** [comment]
> @guozhangwang , I _think_ that reported duplicate scenario is either what is covered in `joinShouldProduceNullsWhenValueHasNonMatchingForeignKey` or the "rapidly changing FK" test I removed. I'm not entirely comfortable with removing the latter test, so maybe I'll add a unit test to exercise this functionality, since it's not possible with TTD.

Sounds good, as long as we do not drop it on the floor :)

**Comment by guozhangwang** [comment]
I do not have further comments, leaving to @mjsax to make a final call and merge.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 7 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Not sure either -- maybe we should test the optimization layer in a more structured way... Also, should be test with and without? No need to update the PR. Just brainstorming in this comment.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: That is weird. Can you investigate? It sound like it exposes a bug.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 11 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: I prefer to avoid noise -- if we don't need to name it, don't name it :) -- but also fine to leave in.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 14 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:87]: Fine with me. It's just an name, and it's descriptive as-is.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 19 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: I think so, especially if we delete a record that did not join and a record that does non-existing in the left table at all, because it should not emit a tombstone.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 21 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: > Sure, we could test all the permutations in one long test, but it's kind of hard to comprehend what such a test is doing after the fact.

Not sure if I can follow. And yes, we should test all permutations according the table in the KIP IMHO.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Actually, now that the tests are so fast, I'll just add an optimization parameter so we test it both ways.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 11 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Ok, I'll take it out.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: Ok, I'll look into it.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 19 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: These are good cases, but I still hesitate to pile too much onto these scenarios (it just makes them hard to follow). I'm adding new tests to cover these cases. Is this ok?

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 21 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Sorry, re-reading that, it's a confusing statement. I just meant that tests that run through super long sequences of mutations are a little hard to understand later on.

Specifically, because the expected output is a function both of the current operation and the prior state the test has built up. So, if you make a change to Streams and an assertion fails halfway through a 500 line test method, it could take you an hour to figure out why the test even failed.

I think it would make more sense to pile all the test cases into one long method in tests where we actually have to run an embedded cluster, since it takes so long to start up and shut down, but for TTD tests, we can afford to have more targeted test methods.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: If you're ok with it, I'll continue this investigation in a separate PR. It would be good to at least get this test improvement merged before figuring out why that cache is misbehaving.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 8 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Sounds good. I've made this a parameterized test, which amounts to the same thing. Not sure what I was thinking...

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: Found in org.apache.kafka.streams.state.internals.CachingKeyValueStore#putAndMaybeForward:

```java
// this is an optimization: if this key did not exist in underlying store and also not in the cache,
// we can skip flushing to downstream as well as writing to underlying store
```

So, it actually is expected that in situations where the join result is actually materialized, we would not see duplicate tombstones (which some of the tests expect).

With this understanding in place, I'm going to leave this line here for now. I'm preparing a follow-on PR to make the `Materialized` parameter optional (and queriability optional even if you specify it). For that PR, I'll have to make specific assertions about this behavior anyway, so we'll have good test coverage there.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 18 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java]: Done (as are your other specific test suggestions above)

**Comment by guozhangwang** [comment]
Merged and cherry-picked to 2.4.

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java (hunk 10 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java:455]: Thanks!

**Comment by mjsax** [comment]
Great work @vvcephei! Thanks for adding even more tests!

**Comment by vvcephei** [comment]
Thanks, @mjsax and @guozhangwang !
